di/_utils/inspect.py

```diff
@@ -47,17 +47,31 @@ def is_gen_callable(call: Any) -> bool:
 
 def get_annotations(call: Callable[..., Any]) -> Dict[str, Any]:
     types_from: Callable[..., Any]
-    if not (
-        inspect.isclass(call) or inspect.isfunction(call) or inspect.ismethod(call)
-    ) and hasattr(call, "__call__"):
-        # callable class
-        types_from = call.__call__  # type: ignore[misc,operator] # accessing __init__ directly
+    if inspect.isclass(call):
+        types_from = call.__init__
+    elif is_callable_class(call):
+        types_from = call.__call__  # type: ignore[operator] # mypy issue #5079
     else:
-        # method
         types_from = call
+
     hints = get_type_hints(types_from, include_extras=True)
-    # for no apparent reason, Annotated[Optional[T]] comes back as Optional[Annotated[Optional[T]]]
-    # so remove the outer Optional if this is the case
+    hints = fix_annotated_optional_type_hints(hints)
+
+    return hints
+
+
+def is_callable_class(call: Callable[..., Any]) -> bool:
+    function_or_method = inspect.isfunction(call) or inspect.ismethod(call)
+    has_call = hasattr(call, "__call__")
+    return not function_or_method and has_call
+
+
+def fix_annotated_optional_type_hints(hints: Dict[str, Any]) -> Dict[str, Any]:
+    """
+    For no apparent reason, `Annotated[Optional[T]]` comes back as `Optional[Annotated[Optional[T]]]`
+    so remove the outer `Optional` if this is the case.
+    """
```

Comment:
This seems really strange. Can you make a MRE? If so we should submit it as a bug to CPython.

Reply:
I only refactored this code (moved the block that handles this logic to function):
https://github.com/adriangb/di/blob/133d696840dca10a2a011f251843332bb0973936/di/_utils/inspect.py#L59-L65

This code is `yours`: https://github.com/adriangb/di/commit/6fce6ca5589c3315225fc348e095f5e90a8567d3

Reply:
Ha okay, I’ll look into it!

Reply:
Here is the issue that the code fixing: https://github.com/python/cpython/issues/90353
I put it in the function docstring.